rfc-126-custom-configuration-for-dependabot.md

```diff
@@ -0,0 +1,246 @@
+# Custom configuration for Dependabot
+
+## Summary
+
+Introduce stricter configuration for Dependabot to reduced number of open PRs, number of deployments and effort required to ensure apps always use the latest version of libraries that we care about.
+
+## Problem
+
+Dependabot was enabled across all our applications in 2018. Due to the number of applications we have and libraries we use, keeping on top of the number of PRs raised by Dependabot has been difficult.
+
+A few solutions have been implemented in the past to make this better:
+
+- [govuk-dependencies] shows the list of open PRs grouped by application, team or gem.
+- [govuk-dependencies] has a Slack bot which automatically messages in each team's channel with a list of open PRs.
+- We [allowed merging Dependabot PRs without requiring two approvals][rfc-103].
+- [Sharing the responsibility for Dependabot PRs across multiple times][dependencies-team].
+
+[govuk-dependencies]: http://govuk-dependencies.herokuapp.com/
+[rfc-103]: https://github.com/alphagov/govuk-rfcs/blob/master/rfc-103-merge-dependabot-pull-requests-with-a-single-review.md
+[dependencies-team]: https://github.com/alphagov/govuk-developer-docs/pull/2168
+
+Although these have made improvements to the process, the number of open PRs remains unmanageable. At the time of writing, there are 135 open PRs, of which there are [8 labelled as security fixes][security-prs], the oldest being 21 days old. Dependabot is currently configured to have at most 5 open PRs per repo, which means this number is not the true number and often new PRs will open the moment existing ones are closed down.
+
+[security-prs]: https://github.com/pulls?q=is%3Apr+author%3Aapp%2Fdependabot-preview+user%3Aalphagov+archived%3Afalse+is%3Aopen+label%3Asecurity
+
+This high number of open PRs is felt particularly hard by the Platform Health team, as they currently [own 55 out of 73 applications][application-owners].
+
+[application-owners]: https://docs.publishing.service.gov.uk/apps/by-team.html
+
+By having so many open PRs, it becomes difficult to be able to prioritise important updates (bug fixes, security fixes) over new versions which simply adds new features (which our apps don't necessarily use anyway). This leads to important updates being left for long periods of time and our desire to close down all the PRs means developers spend time investigating and fixing complex problems without any benefit.
+
+### Statistics
+
+I pulled together [some statistics on our Dependabot usage][dependabot-stats]:
+
+The majority of PRs are merged within a few days (on average 2.5 days). However, the PRs that require manual intervention (for example, failing tests) tend to take a significant amount of time to be merged (on order of months). Because of this talking in terms of averages tends to be less representative of the actual situation, but it can be useful to make comparisons.
```

Comment:
What's the takeaway from this? It seems like a caveat for the subsequent paragraphs. Perhaps we could just provide more info there (mean and max).

Reply:
Yeah, I couldn't decide on the best way of framing these statistics. I thought it would be clearest to use a single number for comparisons between things, but just clarify that the single number is perhaps not the most representative. Another caveat I've just thought of is that mean time to merge ignores the "time between when the PR is raised and work starts to get the tests passing". This wouldn't be solved by including both the mean and the max. 

Ideally, we'd want to show a graph or exclude the outliers in the results (i.e. the one offs that take a really long time and the ones that get merged in within 1 hour always) each time they're used in the RFC, but at the same time I wanted the statistics to supplement the RFC without being the main argument. This is why I settled with explaining a caveat at the top and then continuing to use a single number that easy to understand where it comes from in the body of the RFC.

Reply:
Hmm...I think I'm actually confused about the relevance of timing stats in this RFC, including in the consequences. We don't have any targets for time to merge, and now I'm finding it surprising to see that as a consequence.

Since the time to merge can be due to any and all of volume, variety, and difficulty, I'm not sure we'll see the reduction we're expecting. In any case, it's unclear why "3 days to merge" is much worse than "2 days to merge".

So I'm thinking:

- This is a minor comment on the content here, since it's informational but doesn't seem particularly relevant when you're reading through.
- This is a somewhat minor comment on the importance of mean time to merge as a metric we're looking improve; the count of open PRs seems much easier to work with.

Reply:
Yeah, thinking it through, I'm not sure that using the time to merge metric is particularly useful compared to just a count of the number of PRs. I'm going to reword the RFC to focus on the number of PRs rather than the time. I'll leave my statistics up there if anyone is interested though.

Reply:
I've reworded the RFC to focus on PR numbers in 95a33e35f34a2afdc96c8a73888aa00010d76f16 which I think addresses this!